### PR TITLE
[FEATURE] Support and validate array notation in argument types

### PR DESCRIPTION
**Pull request by pschriner**
Typed arrays are missing. Don't know if that would not be to expensive.

### CODE REVIEW

**Comment by NamelessCoder** [comment]
Hi Patrick - sorry this took a while! We discussed this a while back and came to the conclusion that we would like to support the `ClassName[]` notation - but do not want to support the `ClassOne|ClassTwo` notation. When the type is `ClassName[]` and validation occurs:

* Validate that input is Traversable or array
* When input is empty do nothing
* When input has at least one member, check class of first member only (unless you can determine that checking for example 1,000 members does not decrease performance)

If you could look into this that would be awesome!

**Comment by pschriner** [comment]
I'll look into it, probably no sooner than Friday though :-/

On Fri, Mar 11, 2016 at 2:13 PM, Claus Due <notifications@github.com> wrote:

> Hi Patrick - sorry this took a while! We discussed this a while back and
> came to the conclusion that we would like to support the ClassName[]
> notation - but do not want to support the ClassOne|ClassTwo notation.
> When the type is ClassName[] and validation occurs:
>
>    - Validate that input is Traversable or array
>    - When input is empty do nothing
>    - When input has at least one member, check class of first member only
>    (unless you can determine that checking for example 1,000 members does not
>    decrease performance)
>
> If you could look into this that would be awesome!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/TYPO3Fluid/Fluid/pull/73#issuecomment-195357684>.
>



-- 
"But principally I hate and detest that animal called man, although I
heartily love John, Peter,Thomas, and so forth." - Jonathan Swift

**Comment by mneuhaus** [comment]
Hey @pschriner, could you do that change so we can merge this? :)

**Comment by pschriner** [comment]
Hi @mneuhaus, I'm currently on vacation in the middle of nowhere. Internet access almost nonexistent - I'll hopefully get this done next week when I'm back.

**Comment by mneuhaus** [comment]
Sure, don't worry! i just wanted to follow up :) Happy Holiday ;)

**Comment by NamelessCoder** [comment]
A note for this one for later: we should re-label this a FEATURE since from the perspective of Fluid, it adds a new feature that didn't exist before. The support is originally tied to the TypeConverter of Extbase/Flow which naturally never existed in standalone Fluid - which is why we only have the current basic type validation/conversion.

**Comment by pschriner** [comment]
Sorry for the delay. Hope this is somewhat useful

File: src/Core/ViewHelper/AbstractViewHelper.php [hunk 1]
```diff
@@ -345,6 +329,65 @@ public function validateArguments() {
 	}
 
 	/**
+	 * Check whether the defined type matches the value type
```
Comment (NamelessCoder) [inline src/Core/ViewHelper/AbstractViewHelper.php]: Empty line after description

File: src/Core/ViewHelper/AbstractViewHelper.php [hunk 2]
```diff
@@ -345,6 +329,65 @@ public function validateArguments() {
...
+	 * @param string $type
+	 * @param string $value
+	 * @return boolean
+	 */
+	protected function isValidType($type, $value)
```
Comment (NamelessCoder) [inline src/Core/ViewHelper/AbstractViewHelper.php]: Opening curly on same line as signature (we don't use PSR-2 yet)

File: src/Core/ViewHelper/AbstractViewHelper.php [hunk 3]
```diff
@@ -345,6 +329,65 @@ public function validateArguments() {
...
+	{
+		if ($type === 'object') {
+			if (!is_object($value)) {
+				return false;
```
Comment (NamelessCoder) [inline src/Core/ViewHelper/AbstractViewHelper.php]: Uppercase FALSE, NULL etc.

File: src/Core/ViewHelper/AbstractViewHelper.php [hunk 4]
```diff
@@ -345,6 +329,65 @@ public function validateArguments() {
...
+			}
+		} elseif ($type === 'array' || substr($type, -2) === "[]") {
```
Comment (NamelessCoder) [inline src/Core/ViewHelper/AbstractViewHelper.php]: Single quotes for `'[]'`

File: src/Core/ViewHelper/AbstractViewHelper.php [hunk 5]
```diff
@@ -345,6 +329,65 @@ public function validateArguments() {
...
+			if (!is_array($value) && !$value instanceof \ArrayAccess && !$value instanceof \Traversable && !empty($value)) {
+				return false;
+			} elseif (count($value)>0 && substr($type, -2) === "[]") {
```
Comment (NamelessCoder) [inline src/Core/ViewHelper/AbstractViewHelper.php]: Space between function and comparator/operator.

File: src/Core/ViewHelper/AbstractViewHelper.php [hunk 6]
```diff
@@ -345,6 +329,65 @@ public function validateArguments() {
...
+				$firstElement = $this->getFirstElementOfNonEmpty($value);
+				if ($firstElement == null) {
+					return false;
+				}
+				return $this->isValidType(substr($type, 0, strlen($type)-2), $firstElement);
```
Comment (NamelessCoder) [inline src/Core/ViewHelper/AbstractViewHelper.php]: Space between strlen and -2

File: src/Core/ViewHelper/AbstractViewHelper.php [hunk 7]
```diff
@@ -345,6 +329,65 @@ public function validateArguments() {
...
+		} elseif (is_object($value) && !is_a($value, $type, TRUE)) {
+			return false;
+		}
+		return true;
+	}
+
+	/**
+	 * Return the first element of the given array, ArrayAccess or Traversable
+	 * that is not empty
+	 * @param $value
```
Comment (NamelessCoder) [inline src/Core/ViewHelper/AbstractViewHelper.php]: Empty line after description, missing type of $value in phpdoc

File: src/Core/ViewHelper/AbstractViewHelper.php [hunk 8]
```diff
@@ -345,6 +329,65 @@ public function validateArguments() {
...
+	 */
+	protected function getFirstElementOfNonEmpty($value)
+	{
+		$firstElement = null;
+		if ($value instanceof \ArrayAccess || is_array($value)) {
+			reset($value);
+			$firstElement = current($value);
+		} elseif ($value instanceof \Traversable) {
+			$first = true;
+			foreach ($value as $element) {
```
Comment (NamelessCoder) [inline src/Core/ViewHelper/AbstractViewHelper.php]: Can be replaced by reset and `$value->current()`

**Comment by NamelessCoder** [comment]
Looks like the functionality is in place - now it's just the cosmetic stuff and nitpicking for compactness :)

**Pull request merged**